### ISSUE TITLE
Possible memory leak in Celery integration

### ISSUE THREAD

**Issue by antonpirker** [getsentry/sentry-python#2034]
### How do you use Sentry?

Sentry Saas (sentry.io)

### Version

1.20.0

### Steps to Reproduce

Have a Flask project that has multiple views where one is calling a task (ex called `sometask`) multiple times using the delay method: `sometask.delay(some_msg)`. (In my test case `some_msg` was a string with 10,000 random characters. 

Now load test this project and generate lots of hits to the view (and thus lots of calls to the task)

The memory usage will increase over time:
![plot1 copy 2](https://user-images.githubusercontent.com/202325/234212396-e9156b0b-c89f-4a99-8cba-42d2c240004e.png)

This is the test project I created:
https://github.com/antonpirker/test-celery-memory-leak

### Expected Result

Memory usage stays roughly the same.

### Actual Result

Memory usage increasing.

**Comment by antonpirker** [comment]
Ok, I did a shorter test run (10minutes) and it seems that Sentry is preventing some garbage collection:

Without Sentry:
![without-sentry](https://user-images.githubusercontent.com/202325/234273390-07dda67f-7342-44f8-b391-32fcd4f856e2.png)

With Sentry:
![with-sentry](https://user-images.githubusercontent.com/202325/234273433-de7fa86f-7aa3-4714-9ea3-80fed83428e8.png)

**Comment by samuelhwilliams** [comment]
Hey 👋 Thanks for investigating this after I reported some weird memory issues a while back. This definitely looks promising :) I'll try to have another dig this week and see if I can get a clearer reproduction case on our setup. Out of curiosity was this configured for just error reporting, just transaction tracing, or both?

**Comment by antonpirker** [comment]
I had both, error reporting and tracing. (and around 40% of my requests errored).

If you evaluate this again please make sure to use the lastes Sentry SDK version, because we released some fixes in the mean time. 

All in all we did find a memory leak related to profiling, and also decreased memory usage in combination with redis. 
But we could not find the root cause of the memory usage you reported.

**Comment by jparkrr** [comment]
I'm having a similar issue, except it's with the arq integration instead of celery. (Added in https://github.com/getsentry/sentry-python/pull/1872). It also seems to prevent some garbage collection for me.

Do you think I should create a new issue? Could it be because of the same root cause?

**Comment by antonpirker** [comment]
Yes please create a new issue for that @jparkrr . And it would be great if you could include a sample app to reproduce the possible memory leak

**Comment by williamjamir** [comment]
Does this issue still exist, or has it already been addressed in another PR?

**Comment by antonpirker** [comment]
I just now re-run my tests with the newest SDK (1.45.0) and this is what I got:

With Sentry:
![plot1-with-sentry](https://github.com/getsentry/sentry-python/assets/202325/57ee32fc-cf28-411a-8674-33c88d878b6f)


WIthout Sentry:
![plot1-without-sentry](https://github.com/getsentry/sentry-python/assets/202325/91518ad5-89ff-4b6a-8de1-60b826536a7f)

I do not know what happening at that big bumb. (also had to run my test three times without sentry to get a plot, because other times it crashed vscode that I used to call the scripts from.) So maybe the test is not the best.

Anyhow: the memory usage when using Sentry is stable and not increasing, so no memory leak. 

Without Sentry some memory is freed after some time and with sentry the memory consumption is stable over time. This is not necessarily a bad thing.

**Comment by alchimere** [comment]
@szokeasaurusrex actually I experienced the same behavior on celery integration with SDK 1.45.0, so it may not be fix

**Comment by szokeasaurusrex** [comment]
Hi @alchimere, can you please elaborate on what behavior you experienced? Do you have a chart to show the memory usage over time?

**Comment by rodolfoBee** [comment]
@szokeasaurusrex @sl0thentr0py we have another user on v1.45.0 reporting the same issue in this internal ticket: https://sentry.zendesk.com/agent/tickets/154307

**Comment by szokeasaurusrex** [comment]
Hi @rodolfoBee – unfortunately, version 1.45.0 is no longer supported; we are now on versions 2.x.

The problem should be fixed in the 2.x versions. Since there are no plans to backport the fix to versions 1.x, the user should upgrade to the latest Sentry SDK version (currently 2.29.1) in order to fix the problem. If the user is unable to upgrade to 2.x due to incompatibilities between 1.x and 2.x, then I would recommend that they try disabling the Celery integration to see whether that helps.
